src/packageurl/contrib/purl2url.py

```diff
@@ -237,6 +238,84 @@ def build_hackage_repo_url(purl):
     elif name:
         return f"https://hackage.haskell.org/package/{name}"
 
+@repo_router.route("pkg:golang/.*")
+def build_golang_pkg_go_repo_url(purl):
+    """
+    Return a a download URL from the `purl` string for golang. Due to the non deterministic nature of go package
+    locations this function works in a best effort basis.
+    """
+
+    ##
+    # This function is built using a trial and error method using the golang purl-s I ran across and needed to convert.
+    # To have a more reliable algorithm golang would need to implement a determinaists way to refer to packages or the
+    # method should rely on the active golang proxies (https://proxy.golang.org/ or maybe
+    # https://github.com/gomods/athens)to determine the url.
+
+
+    purl_data = PackageURL.from_string(purl)
+
+    namespace = purl_data.namespace
+    name = purl_data.name
+    version = purl_data.version
+    qualifiers = purl_data.qualifiers
+
+    download_url = qualifiers.get("download_url")
+
+    if download_url:
+        return download_url
+
+    if not (namespace and name and version):
+        return
+    
+    print(f"namespace: {purl_data.namespace}, name {purl_data.name}, version: {purl_data.version}, qualifiers: {purl_data.qualifiers}")
+
+    if "github.com" in purl_data.namespace:
+
+        namespace = purl_data.namespace.split("/")
+        exp = re.compile("v[0-9]+")
+
+        # if the version is a pseudo version and contains several sections separated by - the last section is a git 
+        # commit id what should be referred in the tree of the repo
+        # https://stackoverflow.com/questions/57355929/what-does-incompatible-in-go-mod-mean-will-it-cause-harm
+        if "-" in purl_data.version:
+            version = purl_data.version.split("-")
+            if exp.match(purl_data.name):
```

Comment:
What's the point of this check? Do go package names sometimes have v[0-9]+ in them? Why do we use the name in the download_url in that case?

Reply:
Maybe this check should actually be:
`if len(namespace) == 3:`

Reply:
Oh, perhaps this check is looking for modules with backward incompatiable changes as explained here: https://go.dev/doc/modules/release-workflow

Reply:
Maybe the checks could instead be replaced by:

```
                if len(namespace_parts) == 2:
                    return (
                        f"https://{namespace_parts[0]}/{namespace_parts[1]}/{purl_data.name}"
                        f"/tree/{version}"
                    )
                else:
                    return (
                        f"https://{namespace_parts[0]}/{namespace_parts[1]}/{namespace_parts[2]}"
                        f"/tree/{version}"
                    )
```

I don't see the need for the regular expression check because the backwards incompatiable changes always occur at the 3rd portion of the namespace anyway right?